File: java/name-cache/src/main/java/org/mydotey/namecache/CachedName.java
package org.mydotey.namecache;

/**
 * @author koqizhao
 *
 * Jul 20, 2018
 */
public class CachedName {

    private String _separator;
    private Object[] _nameParts;
    private String _name;

    public CachedName(String separator, Object... nameParts) {
        _separator = separator;
        _nameParts = nameParts;
        generateName();
    }

    protected void generateName() {
        StringBuilder stringBuilder = new StringBuilder();
        for (Object part : _nameParts) {
            if (stringBuilder.length() != 0)
                stringBuilder.append(_separator);
            stringBuilder.append(part);
        }

        _name = stringBuilder.toString();
    }

    public String getSeparator() {
        return _separator;
    }

    public Object[] getNameParts() {
        return _nameParts;
    }

    public String getName() {
        return _name;
    }

    @Override
    public String toString() {
        return _name;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null || obj.getClass() != CachedName.class)
            return false;

        return _name.equals(((CachedName) obj)._name);
    }

    @Override
    public int hashCode() {
        return _name.hashCode();
    }
}

File: dotnet/src/MyDotey.NameCache/CachedName.cs
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Linq;
using System.Text;

namespace MyDotey.NameCache
{
    public class CachedName
    {
        public string Separator { get; private set; }
        public object[] NameParts { get; private set; }
        public string Name { get; private set; }

        public CachedName(string separator, params object[] nameParts)
        {
            Separator = separator;
            NameParts = nameParts;
            GenerateName();
        }

        protected virtual void GenerateName()
        {
            StringBuilder stringBuilder = new StringBuilder();
            foreach (string part in NameParts)
            {
                if (stringBuilder.Length != 0)
                    stringBuilder.Append(Separator);

                stringBuilder.Append(part);
            }

            Name = stringBuilder.ToString();
        }

        public override string ToString()
        {
            return Name;
        }

        public override bool Equals(object obj)
        {
            if (obj == null || obj.GetType() != typeof(CachedName))
                return false;

            return Name.Equals(((CachedName)obj).Name);
        }

        public override int GetHashCode()
        {
            return Name.GetHashCode();
        }

    }
}
